Exploring Ebay Car Sales Data
================
**Synopsis**: We look at a database of used cars listings from _eBay Kleinanzeigen_. The data was scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). The database contains only 50000 data points instead of the original 370000.

**Objective**: The aim of the project is to clean the data and analyze the included used car listings.

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

**Observations**
+ Missing values in the columns: vehicleType, gearbox, model, fuelType, notRepairedDamage.
+ Incorrect type in the columns (correct in parenthesis): dateCrawled (date), price (float), dateCreated (date), lastSeen (date), odometer (float), notRepairedDamage (logical)
+ Incorrect language in the columns: seller, offerType, vehicleType, gearbox, model, fuelType.
+ _name_ column should also be cleaned.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
renamed_columns = ["date_crawled", "name", "seller", "offer_type", "price", "ab_test", "vehicle_type", "registration_year", "gear_box", "power_ps", "model", "odometer", "registration_month", "fuel_type", "brand", "unrepaired_damage", "ad_created", "nr_of_pictures", "postal_code", "last_seen"]

In [7]:
autos.columns = renamed_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

**Changes**

We changed the column names from camel case to snake case for convenience. We also shortened some column names for simplicity.

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

**Observations**
+ There is (mostly) one value in columns: _seller_, _offer_\__type_, _nr_\__of_\__pictures_.
+ Columns that require further investigation: _price_ (many zeroes).
+ Columns with numeric values stored as text: _price_, _odometer_.

In [9]:
print(autos['seller'].value_counts(),"\n")
print("==============")
print(autos['offer_type'].value_counts())

privat        49999
gewerblich        1
Name: seller, dtype: int64 

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64


In [10]:
autos.loc[autos['seller']=="gewerblich"]

date_crawled                                         name  \
7738  2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   

          seller offer_type price  ab_test vehicle_type  registration_year  \
7738  gewerblich    Angebot  $100  control        kombi               2000   

     gear_box  power_ps   model   odometer  registration_month fuel_type  \
7738  manuell         0  megane  150,000km                   8    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
7738  renault               NaN  2016-03-15 00:00:00               0   

      postal_code            last_seen  
7738        65232  2016-04-06 17:15:37

In [11]:
autos.loc[autos['offer_type']=="Gesuch"]

date_crawled                  name  seller offer_type price  \
17541  2016-04-03 15:48:33  Suche_VW_T5_Multivan  privat     Gesuch    $0   

      ab_test vehicle_type  registration_year gear_box  power_ps        model  \
17541    test          bus               2005      NaN         0  transporter   

        odometer  registration_month fuel_type       brand unrepaired_damage  \
17541  150,000km                   0       NaN  volkswagen               NaN   

                ad_created  nr_of_pictures  postal_code            last_seen  
17541  2016-04-03 00:00:00               0        29690  2016-04-05 15:16:06

**Observations**

The rows with unique values for _seller_ and _offer_\__type_ do not seem to show further peculiarities.

Cleaning _price_ and _odometer_ columns
-------------------------------------

In [12]:
autos['price'].value_counts().head()

$0        1421
$500       781
$1,500     734
$2,500     643
$1,200     639
Name: price, dtype: int64

In [13]:
autos['price'] = (autos['price']
                  .str.replace("$","")
                  .str.replace(",","")
                  .astype(float)
                 )

autos['odometer'] = (autos['odometer']
                     .str.replace("km","")
                     .str.replace(",","")
                     .astype(float)
                    )
autos.rename({"odometer": "odometer_km", "price": "price_dollars"}, axis=1, inplace=True)

**_odometer_\__km_ column**

In [14]:
autos['odometer_km'].unique().shape

(13,)

In [15]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [16]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
5000.0        967
Name: odometer_km, dtype: int64

**Observation**

No outliers to remove for the _odometer_\__km_ column.

**_price_ column**

In [17]:
autos['price_dollars'].unique().shape

(2357,)

In [18]:
autos['price_dollars'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64

In [19]:
autos['price_dollars'].value_counts().sort_index(ascending=False).head(15)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: price_dollars, dtype: int64

In [20]:
autos['price_dollars'].value_counts().sort_index(ascending=True).head(25)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
40.0       6
45.0       4
47.0       1
49.0       4
50.0      49
Name: price_dollars, dtype: int64

In [21]:
autos = autos[autos['price_dollars'].between(50,350000)]

In [22]:
autos['price_dollars'].value_counts().sort_index(ascending=False).head()

350000.0    1
345000.0    1
299000.0    1
295000.0    1
265000.0    1
Name: price_dollars, dtype: int64

In [23]:
autos.shape

(48347, 20)

**Changes**

We have removed outliers with regards to the _price_ column from the dataframe. We excluded _price_ values higher than \$350000 and lower than \$50. Values outside the \$50-\$350000 range did not look legitimate. In particular, we have excluded 1421 rows with _price_ equal to 0.

In [24]:
autos['price_dollars'].describe()

count     48347.000000
mean       5915.456099
std        9071.625794
min          50.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price_dollars, dtype: float64

Exploring columns containing dates
-----------------------------------

**_date_\__crawled_ column**

In [44]:
date_crawled_unique = autos['date_crawled'].str[:10].unique().shape[0]

34

In [45]:
date_crawled_total = autos['date_crawled'].str[:10].shape[0]

48347

In [29]:
(autos['date_crawled']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.025358
2016-03-06    0.014065
2016-03-07    0.036031
2016-03-08    0.033260
2016-03-09    0.033053
2016-03-10    0.032225
2016-03-11    0.032556
2016-03-12    0.036983
2016-03-13    0.015658
2016-03-14    0.036672
2016-03-15    0.034294
2016-03-16    0.029495
2016-03-17    0.031584
2016-03-18    0.012907
2016-03-19    0.034749
2016-03-20    0.037789
2016-03-21    0.037293
2016-03-22    0.032949
2016-03-23    0.032267
2016-03-24    0.029433
2016-03-25    0.031501
2016-03-26    0.032267
2016-03-27    0.031088
2016-03-28    0.034894
2016-03-29    0.034108
2016-03-30    0.033715
2016-03-31    0.031832
2016-04-01    0.033735
2016-04-02    0.035535
2016-04-03    0.038575
2016-04-04    0.036528
2016-04-05    0.013051
2016-04-06    0.003165
2016-04-07    0.001386
Name: date_crawled, dtype: float64

In [50]:
print("For a uniform distribution, each value appears with a %1.6f frequency" % (1/date_crawled_unique))

For a uniform distribution, each value appears with a 0.029412 frequency


**Observations**

The distribution looks very roughly uniform, except for some outliers at the beginning and end of the date range.

**_ad_\__created_ column**

In [55]:
ad_created_unique = autos['ad_created'].str[:10].unique().shape[0]
ad_created_unique

76

In [51]:
(autos['ad_created']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033135
2016-03-10    0.031936
2016-03-11    0.032887
2016-03-12    0.036796
2016-03-13    0.017023
2016-03-14    0.035307
2016-03-15    0.034025
2016-03-16    0.029992
2016-03-17    0.031253
2016-03-18    0.013589
2016-03-19    0.033632
2016-03-20    0.037851
2016-03-21 

In [56]:
print("For a uniform distribution, each value appears with a %1.6f frequency" % (1/ad_created_unique))

For a uniform distribution, each value appears with a 0.013158 frequency


In [57]:
(autos['ad_created']
 .str[:10]
 .value_counts(normalize=False, dropna=False)
 .sort_index()
)

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
2015-12-30       1
2016-01-03       1
2016-01-07       1
2016-01-10       2
2016-01-13       1
2016-01-14       1
2016-01-16       1
2016-01-22       1
2016-01-27       3
2016-01-29       1
2016-02-01       1
2016-02-02       2
2016-02-05       2
2016-02-07       1
2016-02-08       1
2016-02-09       1
2016-02-11       1
2016-02-12       2
2016-02-14       2
2016-02-16       1
2016-02-17       1
2016-02-18       2
2016-02-19       3
2016-02-20       2
2016-02-21       3
              ... 
2016-03-09    1602
2016-03-10    1544
2016-03-11    1590
2016-03-12    1779
2016-03-13     823
2016-03-14    1707
2016-03-15    1645
2016-03-16    1450
2016-03-17    1511
2016-03-18     657
2016-03-19    1626
2016-03-20    1830
2016-03-21    1814
2016-03-22    1584
2016-03-23    1552
2016-03-24    1420
2016-03-25    1529
2016-03-26    1563
2016-03-27    1499
2016-03-28    1692
2016-03-29    1647
2016-03-30  

**Observation**

The distribution of _ad_\__created_ does not look uniform. However, we see many outliers corresponding to late 2015/early 2016.

**_last_\__seen_ column**

In [58]:
last_seen_unique = autos['last_seen'].str[:10].unique().shape[0]
last_seen_unique

34

In [59]:
(autos['last_seen']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.001076
2016-03-06    0.004323
2016-03-07    0.005419
2016-03-08    0.007343
2016-03-09    0.009639
2016-03-10    0.010611
2016-03-11    0.012390
2016-03-12    0.023745
2016-03-13    0.008873
2016-03-14    0.012638
2016-03-15    0.015885
2016-03-16    0.016464
2016-03-17    0.028089
2016-03-18    0.007322
2016-03-19    0.015802
2016-03-20    0.020622
2016-03-21    0.020601
2016-03-22    0.021408
2016-03-23    0.018595
2016-03-24    0.019774
2016-03-25    0.019174
2016-03-26    0.016775
2016-03-27    0.015575
2016-03-28    0.020849
2016-03-29    0.022318
2016-03-30    0.024738
2016-03-31    0.023828
2016-04-01    0.022876
2016-04-02    0.024883
2016-04-03    0.025172
2016-04-04    0.024510
2016-04-05    0.124972
2016-04-06    0.221627
2016-04-07    0.132087
Name: last_seen, dtype: float64

**Observations**

The distribution is decidedly not uniform. In fact, most of the values lie between 2016-04-05 and 2016-04-07.

**_registration_\__year_ column**

In [62]:
autos['registration_year'].describe()

count    48347.000000
mean      2004.718845
std         87.786604
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [66]:
autos['registration_year'].value_counts().sort_index()

1000       1
1001       1
1111       1
1800       2
1910       2
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      22
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3120
2001    2632
2002    2478
2003    2695
2004    2699
2005    2921
2006    2670
2007    2274
2008    2212
2009    2081
2010    1589
2011    1620
2012    1309
2013     801
2014     662
2015     384
2016    1207
2017    1385
2018     468
2019       2
2800       1
4100       1
4500       1
4800       1
5000       3
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

**Observations**

There are some outliers, namely years larger than 2016 (the scraping year) and lower than 1900. Realistically, cars registered earlier than this would not be sold on eBay.

In [67]:
autos = autos[autos['registration_year'].between(1900,2016)]

In [68]:
autos.shape

(46473, 20)

In [70]:
autos['registration_year'].value_counts(normalize=True).sort_index()

1910    0.000043
1927    0.000022
1929    0.000022
1931    0.000022
1934    0.000043
1937    0.000086
1938    0.000022
1939    0.000022
1941    0.000043
1943    0.000022
1948    0.000022
1950    0.000022
1951    0.000043
1952    0.000022
1953    0.000022
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000473
1961    0.000129
1962    0.000086
1963    0.000172
1964    0.000258
1965    0.000366
1966    0.000473
1967    0.000559
1968    0.000559
          ...   
1987    0.001549
1988    0.002883
1989    0.003680
1990    0.007230
1991    0.007295
1992    0.007919
1993    0.009124
1994    0.013535
1995    0.025929
1996    0.029350
1997    0.041745
1998    0.050589
1999    0.062058
2000    0.067136
2001    0.056635
2002    0.053321
2003    0.057991
2004    0.058077
2005    0.062854
2006    0.057453
2007    0.048932
2008    0.047598
2009    0.044779
2010    0.034192
2011    0.034859
2012    0.028167
2013    0.017236
2014    0.0142

**Observations**

The vast majority of the cars listed were registered after 1994.
The mode is 2000 and indeed most of the mass lies in the range 1997-2009.

Exploring the brand column
--------------------------

In [78]:
autos['brand'].unique().shape

(40,)

In [120]:
autos['brand'].value_counts().head(10)

volkswagen       9821
bmw              5110
opel             4987
mercedes_benz    4485
audi             4025
ford             3252
renault          2193
peugeot          1390
fiat             1195
seat              848
Name: brand, dtype: int64

In [118]:
brands = autos['brand'].value_counts(normalize=True) > 0.05
top_brands = brands[brands == True].index
top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

We will begin by aggregating on the brands with more than 5% of the total number of postings.

In [119]:
aggregate_price = {}
for b in top_brands:
    price = autos.loc[autos['brand'] == b,"price_dollars"].mean()
    aggregate_price[b] = price
    
aggregate_price

{'audi': 9373.771428571428,
 'bmw': 8376.805088062622,
 'ford': 3762.1165436654364,
 'mercedes_benz': 8663.059308807135,
 'opel': 2996.0850210547424,
 'volkswagen': 5424.9185419000105}

**Observations**

We see that cars that are sold the most frequently are not necessarily the cheapest. Moreover, there seems to be three categories:
+ Audi, BMW, Mercedes Benz are the most expensive.
+ Ford, Opel are the cheapest.
+ Volkswagen is in between.

In [123]:
aggregate_mileage = {}
for b in top_brands:
    mileage = autos.loc[autos['brand'] == b,"odometer_km"].mean()
    aggregate_mileage[b] = mileage
    
aggregate_mileage

{'audi': 129254.65838509316,
 'bmw': 132651.66340508807,
 'ford': 124295.81795817958,
 'mercedes_benz': 130923.07692307692,
 'opel': 129440.54541808703,
 'volkswagen': 128771.00091640363}

In [126]:
aggr_price_series = pd.Series(aggregate_price)
aggr_mileage_series = pd.Series(aggregate_mileage)

price_mileage_df = pd.DataFrame(aggr_price_series, columns = ['mean_price'])
price_mileage_df["mean_mileage"] = aggr_mileage_series

price_mileage_df

mean_price   mean_mileage
audi           9373.771429  129254.658385
bmw            8376.805088  132651.663405
ford           3762.116544  124295.817958
mercedes_benz  8663.059309  130923.076923
opel           2996.085021  129440.545418
volkswagen     5424.918542  128771.000916

**Observations**

We see that the mean mileage is roughly constant across different brands. We conclude that the differences in _mean_\__price_ cannot be explained by differences in mileage.